### PR TITLE
Appeals 35131 duplicate veteran checker refactor

### PR DESCRIPTION
**Pull request by nicorithner-bah**
Resolves [Automate Duplicate Veteran Checker](https://jira.devops.va.gov/browse/APPEALS-35142)
Resolves [Test Duplicate Veteran Checker](https://jira.devops.va.gov/browse/APPEALS-37013)

## Description
The original file DuplicateVeteranChecker had a lot of redundancies and it didn't have tests associated to it.

## Acceptance Criteria
- [x] Redundancies are removed
- [x] There tests for methods and they pass

## Testing Plan

1. Go to [Jira Issue/Test Plan Link](https://jira.devops.va.gov/browse/JIRA-12345) or list them below

- [ ] For feature branches merging into master: Was this deployed to UAT?

#### Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [x] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

**Comment by ronwabVa** [comment]
Hey @nicorithner-bah  please revert the changes you made to `war_room.rb` and `duplicate_veteran_checker.rb`. These files are needed to manually fix duplicate vets by I believe tier 4 or tier 2. Once its done we can start reviewing all the other changes.

**Comment by nicorithner-bah** [comment]
> Hey @nicorithner-bah please revert the changes you made to `war_room.rb` and `duplicate_veteran_checker.rb`. These files are needed to manually fix duplicate vets by I believe tier 4 or tier 2. Once its done we can start reviewing all the other changes.

Thank you for the heads up @ronwabVa Ron! I brought it back now ✅

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+  end
+
+  ##
+  # Gathers relations associated with duplicate file number and correct file number
+  # stores the relations and their count in a hash and logs results to the console.
+  ##
+  def log_and_get_relations_count
+    correct_file_number = find_bgs_file_number_with_ssn
+
+    check_only_one_vet_has_bad_file_number(@duplicate_veteran_file_number)
```
Comment (davywentwortht2it) [inline lib/helpers/duplicate_veteran_fixer.rb:54]: if there ISN'T only one vet with a bad file number, should we short-circuit?

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+  ##
+  def update_failure_log(message)
+    @failure_log.push(message)
+    puts(message)
+  end
+
+  ##
+  # Sets duplicate pair
+  # Grab first of the dupilicates and check if the duplicate veteran}
+  # This return an ssn. What am I supposed to do with i??
```
Comment (davywentwortht2it) [inline lib/helpers/duplicate_veteran_fixer.rb:409]: this comment needs updating.

File: spec/lib/helpers/duplicate_veteran_fixer_spec.rb [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    end
+  end
+
+  context "update relations and log" do
+    it "changes legacy to use vbms id" do
+      expect(legacy_appeal.case_record[:bfcorlid]).to eq(dup_vbms_id)
+      expect(legacy_appeal.case_record.folder[:titrnum]).to eq(dup_vbms_id)
+      expect(legacy_appeal.case_record.correspondent[:slogid]).to eq(dup_vbms_id)
+
+      # subject.change_legacy_to_use_vbms_id(legacy_appeal, vbms_id)
```
Comment (davywentwortht2it) [inline spec/lib/helpers/duplicate_veteran_fixer_spec.rb:224]: remove comment.

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+    las = @duplicate_relations[:las][:relations].first
+
+    change_legacy_to_use_vbms_id(las, vbms_id) if !las.blank?
+
+    update_file_numbers(bgs_file_number, vbms_id, @duplicate_veteran_file_number)
+
+    if @updated_relations_count_log.length != 1
+      puts("There were differences in duplicate relations and update relations.")
+      send_log(@updated_relations_count_log)
+      puts("Stoping script here. Need manual intervention")
```
Comment (davywentwortht2it) [inline lib/helpers/duplicate_veteran_fixer.rb:87]: spelling. 'stopping'

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+    end
+
+    send_log(@updated_relations_count_log)
+
+    if remaining_duplicates?(@duplicate_veteran_file_number)
+
```
Comment (davywentwortht2it) [inline lib/helpers/duplicate_veteran_fixer.rb:93]: remove blank line.

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+  end
+
+  ##
+  # Updates relations associated with duplicate veteran record
+  # to use BGS - correct - file number.
+  # Logs updated relations count
+  # Check for remaining duplicate relations
+  # Logs remaining relations count
+  ##
+  def update_relations
```
Comment (davywentwortht2it) [inline lib/helpers/duplicate_veteran_fixer.rb:74]: can this method eventually be moved to `private`?

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 1 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:54]: @ronwabVa what do you think? I agree with Davy but we talked about no having 'fail interrupt' in this file.

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 5 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:93]: ✅

File: spec/lib/helpers/duplicate_veteran_fixer_spec.rb (hunk 3 above)
Comment (nicorithner-bah) [reply spec/lib/helpers/duplicate_veteran_fixer_spec.rb:224]: ✅

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+    @duplicate_veterans ||= Veteran.where("ssn = ? or participant_id = ?", duplicate_veteran.ssn,
+                                          duplicate_veteran.participant_id)
+  end
+
+  def vet_ssn
+    duplicate_veteran.ssn
+  end
+
+  def find_bgs_file_number_with_ssn
+    BGSService.new.fetch_file_number_by_ssn(vet_ssn)
```
[the remark this answers is not in the document]
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:144]: ✅

File: app/jobs/duplicate_veteran_fixer_job.rb [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    ActiveRecord::Base.transaction do
+      object_type.clear_error!
+    rescue StandardError => error
+      log_error(error)
+      @stuck_job_report_service.append_errors(object_type.class.name, object_type.id, error)
+    end
+  end
+
+  def records_with_errors
+    SupplementalClaim.where("establishment_error ILIKE ?", "%#{error_text}%")
```
[the remark this answers is not in the document]
Comment (nicorithner-bah) [reply app/jobs/duplicate_veteran_fixer_job.rb:102]: ✅

File: app/jobs/duplicate_veteran_fixer_job.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  # @overload Gets list of relations and logs count as a string
+  # @param [string] duplicate veteran file number
+  # :reek:UtilityFunction
+  ##
+  def get_relations_and_log_count(duplicate_veteran_file_number)
+    dvf = DuplicateVeteranFixer.new(duplicate_veteran_file_number)
+    dvf.log_and_get_relations_count
+  end
+
+  def perform
```
[the remark this answers is not in the document]
Comment (nicorithner-bah) [reply app/jobs/duplicate_veteran_fixer_job.rb:51]: ✅ Move perform to top of file

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 2 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:409]: ✅

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 1 above)
Comment (ronwabVa) [reply lib/helpers/duplicate_veteran_fixer.rb:54]: @nicorithner-bah Can we use return ? I thought that was what we agreed to do

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 6 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:74]: ✅

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 1 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:54]: added implicit return

File: spec/lib/helpers/duplicate_veteran_fixer_spec.rb [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+RSpec.feature DuplicateVeteranFixer do
+  include_context "veterans"
+
+  subject { DuplicateVeteranFixer.new(duplicate_file_number) }
+
+  context "run_remediation fixes veteran records duplication" do
+    describe "There is a pair of duplicates" do
+      it "starts with a pair of duplicate veteran records with same ssn and different file number" do
+        duplicate_vet_fixer
+        duplicate_veterans = duplicate_vet_fixer.send(:duplicate_veterans)
```
Comment (ronwabVa) [inline spec/lib/helpers/duplicate_veteran_fixer_spec.rb:16]: We talked about not testing the private methods this way.

File: spec/lib/helpers/duplicate_veteran_fixer_spec.rb [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        allow(duplicate_vet_fixer).to receive(:duplicate_veterans).and_return([dup_veteran1, dup_veteran2])
+
+        expect(Veteran.exists?(dup_veteran1.id)).to be true
+        duplicate_vet_fixer.run_remediation
+        expect(Veteran.exists?(dup_veteran1.id)).to be false
+      end
+    end
+  end
+
+  context "get relations count and log" do
```
Comment (ronwabVa) [inline spec/lib/helpers/duplicate_veteran_fixer_spec.rb:89]: Can we get these counts from calling `run_remediation` ? That way we will be testing the actual code flow instead of testing the private methods

File: spec/lib/helpers/shared/veterans_context.rb [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+  let!(:dup_veteran2) do
+    create(:veteran, file_number: dup_pair_vet_number2, ssn: ssn2, participant_id: participant_id_2)
+  end
+
+  # --- Instanciate duplicate veteran fixer
+  let(:duplicate_vet_fixer) { DuplicateVeteranFixer.new(dup_pair_vet_number1) }
+
+  # --- Set variables
+  let!(:bgs) { dup_veteran2.file_number }
```
Comment (ronwabVa) [inline spec/lib/helpers/shared/veterans_context.rb:40]: `bgs` is the same as `dup_pair_vet_number2`  No need to call `dup_veteran2.file_number`

File: spec/lib/helpers/shared/veterans_context.rb [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+  let(:remaining_message_after_update) { "No remaining relations with duplicate veteran file number" }
+
+  # -----------------------------------------------------------------------------------
+  # -- Lists of relations with duplicate file number.
+  # -----------------------------------------------------------------------------------
+
+  let!(:appeals) { create_list(:appeal, 2, veteran_file_number: duplicate_file_number) }
+
+  let!(:legacy_appeal) do
```
Comment (ronwabVa) [inline spec/lib/helpers/shared/veterans_context.rb:134]: Please stick to one naming convention.. Its hard to follow your thought process So you have `v1_legacy_appeal` and now you have `legacy_appeal` is legacy_appeal for the duplicate vet? Do we ever create a legacy appeal with `v2_file_number`?

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+    is_vi: { class: Intake, name: " Intakes related by veteran id" }, # by veteran id
+    res: { class: RampElection, name: " RampElection" },
+    rrs: { class: RampRefiling, name: " RampRefiling" },
+    scs: { class: SupplementalClaim, name: " SupplementalClaims" }
+  }.freeze
+
+  ##
+  # Runs methods to find duplicates, update relations, and delete extra veteran record.
+  ##
+  def run_remediation
```
Comment (AdamShawBAH) [inline lib/helpers/duplicate_veteran_fixer.rb:41]: general additions of error handling. Will mark different methods where it could be helpful to add.

for here in terms of issues that might arise during the execution of the remediation process, like failures in logging or updating relations

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+    log_and_get_relations_count
+    update_relations
+    delete_duplicate_veteran
+  end
+
+  ##
+  # Gathers relations associated with duplicate file number and correct file number
+  # stores the relations and their count in a hash and logs results to the console.
+  ##
+  def log_and_get_relations_count
```
Comment (AdamShawBAH) [inline lib/helpers/duplicate_veteran_fixer.rb:51]: here for handling for cases where the correct file number cannot be obtained or if there is an issue with gathering relations associated with duplicate and correct file numbers

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 16]
```diff
@@ -0,0 +1,531 @@
...
+  private
+
+  ##
+  # Updates relations associated with duplicate veteran record
+  # to use BGS - correct - file number.
+  # Logs updated relations count
+  # Check for remaining duplicate relations
+  # Logs remaining relations count
+  ##
+  def update_relations
```
Comment (AdamShawBAH) [inline lib/helpers/duplicate_veteran_fixer.rb:89]: here for handling errors that might occur during the update of relations

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 17]
```diff
@@ -0,0 +1,531 @@
...
+
+      message = "Duplicate veteran still has associated records. Can not delete until resolved:\n"
+      puts("#{message} #{@remaining_relations_count_log.join}")
+    end
+  end
+
+  ##
+  # Deletes duplicate veteran
+  ##
+  def delete_duplicate_veteran
```
Comment (AdamShawBAH) [inline lib/helpers/duplicate_veteran_fixer.rb:118]: here for handling for potential issues during the deletion of the duplicate veteran, like failure to find or destroy the veteran record

i know you have the puts, but might be helpful to have the error log with handling

File: spec/lib/helpers/duplicate_veteran_fixer_spec.rb [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+require "helpers/duplicate_veteran_fixer"
+require "./spec/lib/helpers/shared/veterans_context"
+
+# Tests it fixes veteran duplication
+RSpec.feature DuplicateVeteranFixer do
+  include_context "veterans"
+
+  subject { DuplicateVeteranFixer.new(duplicate_file_number) }
+
+  context "run_remediation fixes veteran records duplication" do
```
[the remark this answers is not in the document]
Comment (nicorithner-bah) [reply spec/lib/helpers/duplicate_veteran_fixer_spec.rb:12]: It's testing run remediation directly.

File: spec/lib/helpers/shared/veterans_context.rb (hunk 12 above)
Comment (nicorithner-bah) [reply spec/lib/helpers/shared/veterans_context.rb:40]: I add them to help with readability

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 15 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:51]: ✅

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 16 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:89]: ✅

File: lib/helpers/duplicate_veteran_fixer.rb (hunk 17 above)
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:111]: ✅

File: lib/helpers/duplicate_veteran_fixer.rb [hunk 19]
```diff
@@ -0,0 +1,531 @@
...
+
+  def vet_ssn
+    duplicate_veteran.ssn
+  end
+
+  def correct_file_number_by_ssn
+    @correct_file_number_by_ssn ||= BGSService.new.fetch_file_number_by_ssn(vet_ssn)
+  end
+
+  def find_vbms_id
```
[the remark this answers is not in the document]
Comment (nicorithner-bah) [reply lib/helpers/duplicate_veteran_fixer.rb:151]: ✅

**Comment by nicorithner-bah** [comment]
Screenshot 2024-01-30 at 7 23 44 AM

Screenshot 2024-01-30 at 7 25 28 AM

Still need to add tests for these two logs. It may require a small refactor of method `pair_is_duplicate?(dp_vets)`
